src/simple.sml

```diff
@@ -50,6 +51,65 @@ struct
            List.find (fn (su,r) => r = r') (map #1 (List.nth (#hands s,i)))
        | _ => NONE
 
+  (* Gives the clued rank of a card, returns NONE if no number clue has been given *)
+  fun cluedRank (is : info list) : int option =
+    case List.find (fn i => case i of IsRank _ => true | _ => false) is of
+      SOME (IsRank r) => SOME r
+     | _ => NONE
+
+  (* Gives the list of negative rank information of a card *)
+  fun cluedNotRank (is : info list) : int list =
+    List.mapPartial (fn i => case i of NotRank r => SOME r | _ => NONE) is
+
+  (* Gives the list of negative color information of a card *)
+  fun cluedNotSuit (is : info list) : suit list =
+    List.mapPartial (fn i => case i of NotSuit su => SOME su | _ => NONE) is
+
+  (* Returns whether a card is clued to be not rainbow *)
+  fun isNotRainbow (is : info list) : bool =
+    List.exists (fn i => case i of NotSuit _ => true | _ => false) is
+
+  (* Gives the clued suit, returns Rainbow if 2 different suit clues have been given *)
+  fun cluedSuit (is : info list) : suit option =
+  let
+    val sus = List.mapPartial (fn i => case i of IsSuit su => SOME su | _ => NONE) is
+  in
+    if null sus then NONE
+    else if List.exists (fn j => j <> hd sus) (tl sus)
+    then SOME Rainbow
+    else SOME (hd sus)
+  end
+
+  (* Returns the list of possible suits a card can have *)
+  fun possibleSuits (is : info list) : suit list =
+    case cluedSuit is of
+	 SOME Rainbow => [Rainbow]
+       | SOME su => if isNotRainbow is then [su] else [su, Rainbow]
+       | NONE =>
+	 let
+	     val lsu = cluedNotSuit is
+	 in
+	     (List.filter (fn su => List.all (fn su' => su <> su') lsu)
+			  normalSuits) @ (if null lsu then [Rainbow] else [])
```

Comment:
I think the first part of this is `List.filter (fn su => not (Util.elem lsu su)) normalSuits`.